README.md
# SuperRandom

* [VERSION 3.2.230213](https://github.com/carlosjhr64/super_random/releases)
* [github](https://github.com/carlosjhr64/super_random)
* [rubygems](https://rubygems.org/gems/super_random)

## DESCRIPTION:

You can't get more random than random, but you can try really, really, really hard.

`SuperRandom` combines sources of entropy to generate super-random bytes!

## INSTALL:

    $ gem install super_random

## SYNOPSIS:
```ruby
require 'super_random'
SuperRandom::VERSION              #~> ^3.\d+.\d+$
SuperRandom::DEFAULT_SOURCES      #~> www.random.org
super_random = SuperRandom.new
super_random.sources              #~> www.random.org
super_random.bytes                #~> ^\[\d+(, \d+){63}\]$
# The source_count attribute gives the number of sources successfully used.
super_random.source_count         #=> 1
# The byte_count attribute gives the total bytes digested from sources.
super_random.byte_count           #=> 210
super_random.hexadecimal          #~> ^\h{128}$
super_random.random_number(100.0) #~> ^\d{1,2}\.\d+$
super_random.random_number(100)   #~> ^\d{1,2}$
# Because of a 1 minute rate limit, subsequent source counts are zero:
super_random.source_count         #=> 0
super_random.byte_count           #=> 0
# Snapshots from your webcam can be a good entropy source:
super_random = SuperRandom.new('/var/lib/motion/lastsnap.jpg')
super_random.sources              #=> ["/var/lib/motion/lastsnap.jpg"]
# Dice!
d6 = super_random.dice(6)
d6.roll                           #~> ^[123456]$
```
## METHODOLOGY:

`SuperRandom` uses `OpenUri` to read your sources, which
can be http, https, or ftp URLs.
`Digest::SHA2.new(512)` is used to digest your sources.
Finally, `SecureRandom.bytes` are fed to the digest as a fail safe.
This generates the final 64 random bytes.

## SOURCES:

I could only find one good source expressly for this purpose, `www.random.org`.
Another good one is `qrng.anu.edu.au`, but you'll need an API key.
Consider using:

* Snapshots from your webcam
* List of market spot prices
* Weather forecasts
* News or micro-logging feeds

Be very nice about your calls,
specially when you're not using the source as intended.
`SuperRandom` enforces a one minute rate limit in the use of sources.
Here are ways to set custom sources:
```ruby
# Sources specified in the constructor: 
super_random = SuperRandom.new('/var/lib/motion/lastsnap.jpg', 'https://wttr.in')
super_random.sources
#=> ["/var/lib/motion/lastsnap.jpg", "https://wttr.in"]

# Sources appended on the instance: 
super_random.sources.append 'https://text.npr.org'
super_random.sources
#=> ["/var/lib/motion/lastsnap.jpg", "https://wttr.in", "https://text.npr.org"]

# Sources appended to the DEFAULT_SOURCES
SuperRandom::DEFAULT_SOURCES.append 'https://coinmarketcap.com'
super_random = SuperRandom.new
super_random.sources
#~> random.org.*coinmarketcap.com
```
## LICENSE:

(The MIT License)

Copyright (c) 2023 CarlosJHR64

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.0.0 / 2014-12-04 11:08:21 -0800
=== 0.0.0 / 2014-12-04 12:13:15 -0800
93065b089944456e41307fb95063c529  pkg/super_random-0.0.0.gem
=== 0.0.1 / 2014-12-04 14:29:19 -0800
aa50c24135172cd90c9600bc65c5f299  pkg/super_random-0.0.1.gem
=== 0.0.2 / 2014-12-07 12:45:56 -0800
23c2646dc46071f8a9ff616a3a93b85b  pkg/super_random-0.0.2.gem
=== 0.0.3 / 2015-12-16 16:03:26 -0800
3d310b89caf1fac5237def5ae7328b16  pkg/super_random-0.0.3.gem
=== 0.1.0 / 2017-01-07 08:40:16 -0800
0db7d72865bbaf2836178fed015fdd92  pkg/super_random-0.1.0.gem
=== 0.2.0 / 2017-06-13 08:58:42 -0700
4b41e844f32bcb661227977cedd01912  pkg/super_random-0.2.0.gem
=== 1.0.0 / 2017-08-22 14:11:32 -0700
1b89e83eefe2cd52dc6b583f1192148e  pkg/super_random-1.0.0.gem
=== 1.1.0 / 2019-12-18 17:06:03 -0800
42e7123fd6e7f00f3008341e7dc0d1e5  pkg/super_random-1.1.0.gem
# 1.2.200123 / 2020-01-23 17:40:14 -0800
ced371a20480cbb400d2e471f1a304bc  pkg/super_random-1.2.200123.gem
# 2.0.210126 / 2021-01-26 12:00:22 -0800
746a94ff6ffae77c00e1001a035e9923  pkg/super_random-2.0.210126.gem
# 3.0.230110 / 2023-01-10 16:44:59 -0800
a3212499d92ccad0c4d2937268fa7259  pkg/super_random-3.0.230110.gem
# 3.0.230111 / 2023-01-11 09:48:17 -0800
7945752416889f4dd270beed77bfdb7e  pkg/super_random-3.0.230111.gem
# 3.0.230112 / 2023-01-12 06:08:49 -0800
85e62ddb62e50e3c9bab3a429e0b50e4  pkg/super_random-3.0.230112.gem
# 3.0.230113 / 2023-01-13 07:44:18 -0800
dd2aa8a7e49f5dcaa77a97a3c9485522  pkg/super_random-3.0.230113.gem
# 3.1.230114 / 2023-01-14 14:37:55 -0800
7c0e4e07caf4e9d053043bec5975626b  pkg/super_random-3.1.230114.gem
# 3.2.230116 / 2023-01-16 08:32:58 -0800
ba62e9adb5a35d52be895fdf8f1f0a11  pkg/super_random-3.2.230116.gem
# 3.2.230118 / 2023-01-18 04:50:01 -0800
e64e14560a3982d61748328740c88c62  pkg/super_random-3.2.230118.gem
# 3.2.230213 / 2023-02-13 13:27:31 -0800
ef572e34cda7c6ff0fc902a0f688dacb  pkg/super_random-3.2.230213.gem


lib/super_random/services.rb
class SuperRandom
  def self.services
    [:quantum, :atmospheric, :hotbits]
  end

  # https://qrng.anu.edu.au/
  # https://qrng.anu.edu.au/contact/api-documentation/
  def self.quantum(n)
    s = Net::HTTP.get(URI(
      "https://qrng.anu.edu.au/API/jsonI.php?length=#{n}&type=uint8"))
    a = JSON.parse(s)['data']
    raise unless a.is_a?(Array) and a.length==n
    raise unless a.all?{|i| i.is_a?(Integer) and i.between?(0,255)}
    return a
  rescue StandardError
    warn "quantum (qrng.anu.edu.au) failed."
    return nil
  end

  # https://www.random.org/
  # https://www.random.org/integers/
  def self.atmospheric(n)
    s = Net::HTTP.get(URI(
      "https://www.random.org/integers/?num=#{n}&min=0&max=255&col=1&base=10&format=plain&rnd=new"))
    a = s.strip.split(/\s+/).map{|j|j.to_i}
    raise unless a.length==n
    raise unless a.all?{|i| i.between?(0,255)}
    return a
  rescue StandardError
    warn "atmospheric (www.random.org) failed."
    return nil
  end

  # https://www.fourmilab.ch/hotbits/
  def self.hotbits(n, k='Pseudorandom')
    s = Net::HTTP.get(URI(
      "https://www.fourmilab.ch/cgi-bin/Hotbits.api?nbytes=#{n}&fmt=bin&apikey=#{k}"))
    a = s.bytes
    raise unless a.length==n
    return a
  rescue StandardError
    warn "hotbits (www.fourmilab.ch) failed."
    return nil
  end
end


lib/super_random/generator.rb
class SuperRandom
  DEFAULT_BYTES = 32

  attr_accessor :first_timeout, :second_timeout, :nevermind
  attr_reader :randomness, :services

  def initialize
    @first_timeout = 3
    @second_timeout = 6
    @nevermind = true
    @randomness = 0.0
    @services = 0
  end

  def bytes(n=DEFAULT_BYTES)
    @randomness = 0.0
    @services = 0

    m = SuperRandom.services
    a = Array.new m.length
    t = Array.new m.length
    m.each_with_index do |k,i|
      t[i] = Thread.new{ a[i] = SuperRandom.send(k, n) }
    end

    begin
      Timeout.timeout(@first_timeout) do
        # Initially, would like to get them all.
        t.each{_1.join}
      end
    rescue Timeout::Error
      begin
        Timeout.timeout(@second_timeout) do
          # But at this point,
          # would like to get at least one.
          while a.all?{_1.nil?} and t.any?{_1.alive?}
            Thread.pass
          end
        end
      rescue Timeout::Error
        # If we don't care that we got nothing, go on.
        raise $! unless @nevermind
      end
    end

    r = Array.new n
    n.times{|i| r[i] = SecureRandom.random_number(256)}

    a.each do |b|
      next if b.nil?
      l = b.length
      @randomness += l.to_f/n.to_f
      @services += 1
      n.times{|i|r[i]=(r[i]+b[i%l])%256}
    end

    return r
  end

  def hexadecimal(n=DEFAULT_BYTES)
    bytes(n).map{|i|i.to_s(16).rjust(2,'0')}.join
  end

  def random_number(scale=1.0, minbytes=6, maxbytes=[minbytes,DEFAULT_BYTES].max)
    case scale
    when Float
      div = minbytes.times.inject(''){|s,i| s+'FF'}.to_i(16)
      den = hexadecimal(minbytes).to_i(16)
      return  scale * den.to_f / div.to_f
    when Integer
      n = n0 = Math.log(scale, 256).ceil
      e = e0 = 256**n
      r = r0 = e0 % scale
      while r > 0
        n0 += 1
        e0 *= 256
        r0 = e0 % scale
        if r0 <= r
          # break if repeating pattern with big enough integer
          break if r0 == r and n0 > minbytes
          r,n,e = r0,n0,e0
        end
        break if n0 >= maxbytes
      end
      max = (e/scale)*scale
      loop do
        number = hexadecimal(n).to_i(16)
        return number % scale if number < max
        # On a relatively small chance that we're above max...
        if @nevermind
          warn "using SecureRandom.random_number(#{scale})"
          return SecureRandom.random_number(scale)
        end
      end
    end
    raise "rand(scale Integer|Float)"
  end
  alias rand random_number
end


lib/super_random.rb
require 'timeout'
require 'securerandom'
require 'open-uri'
require 'digest'
# Requires:
#`ruby`

class SuperRandom
  VERSION = '3.2.230213'
  DEFAULT_SOURCES = [
    'https://www.random.org/strings/?' \
    'num=10&len=20&digits=on&upperalpha=on&loweralpha=on&unique=on&' \
    'format=plain&rnd=new'
  ]
  MUTEX = Mutex.new
  DIV = 128.times.inject(''){|h,_|h<<'f'}.to_i(16)
  RATE_LIMIT = 60 # One minute, be nice!
  @@last_time = Time.now.to_i - RATE_LIMIT - 1

  attr_accessor :first_timeout, :second_timeout, :nevermind
  attr_reader   :sources, :source_count, :byte_count

  def initialize(*sources)
    @sources = sources.empty? ? DEFAULT_SOURCES.dup : sources
    @byte_count,@source_count = 0,0
    @first_timeout,@second_timeout,@nevermind = 3,3,true
  end

  # Returns 64 random bytes(at least from SecureRandom's)
  def bytes
    @byte_count,@source_count = 0,0
    _do_updates if Time.now.to_i - @@last_time > RATE_LIMIT
    _get_bytes
  end

  def hexadecimal
    bytes.map{_1.to_s(16).rjust(2,'0')}.join
  end

  def integer
    hexadecimal.to_i(16)
  end

  def xy
    hex = hexadecimal
    x = hex[0...64].to_i(16)
    y = hex[64..128].to_i(16)
    [x,y]
  end

  def integer2
    # An alternate way to generate an integer
    x,y = xy
    # I think this is Binomial(Gaussian in the limit) around zero?
    x - y
  end

  def float
    Rational(integer, DIV).to_f
  end

  def float2
    # An alternate way to generate a float...
    x,y = xy
    # ...but what distribution is this?
    Rational(x+1,y+1).to_f
  end

  def random_number(scale=1.0)
    case scale
    when Float
      return scale * float
    when Integer
      return ((integer + SecureRandom.random_number(scale)) % scale)
    end
    raise 'rand(scale Integer|Float)'
  end
  alias rand random_number

  class Dice
    private def set_big
      @big = @rng.integer + SecureRandom.random_number(@sides)
    end
    def initialize(sides, minimum:1, rng:SuperRandom.new)
      @sides,@minimum,@rng = sides,minimum,rng
      set_big
    end
    def roll
      @big,roll = @big.divmod(@sides)
      roll+@minimum
    ensure
      set_big unless @big.positive?
    end
  end
  def dice(sides, minimum:1)
    Dice.new(sides, minimum:minimum, rng:self)
  end

  private

  SHASUM = Digest::SHA2.new(512)
  def _get_bytes
    MUTEX.synchronize do
      SHASUM.update SecureRandom.bytes(64)
      SHASUM.digest.chars.map(&:ord)
    end
  ensure
    MUTEX.synchronize{SHASUM.update SecureRandom.bytes(64)}
  end

  # source in @sources may be anything URI.open can handle:
  def _update_with(source)
    URI.open(source) do |tmp|
      MUTEX.synchronize do
        tmp.each_line do |line|
          SHASUM.update line
          @byte_count += line.bytesize
        end
        @source_count += 1
      end
    end
  rescue
    warn $!.message
  end

  def _do_updates
    @@last_time = Time.now.to_i
    threads = @sources.inject([]){|t,s|t.push Thread.new{_update_with s}}
    begin
      # Initially, would like to get them all.
      Timeout.timeout(@first_timeout){threads.each(&:join)}
    rescue Timeout::Error
      begin
        Timeout.timeout(@second_timeout) do
          # But at this point,
          # would like to get at least one.
          Thread.pass while @source_count.zero? && threads.any?(&:alive?)
        end
      rescue Timeout::Error
        MUTEX.synchronize{threads.each(&:exit)} # Exit each thread
        raise $! unless @nevermind # Go on if never-minding
      end
    end
  end
end
